In [1]:
import pandas as pd
from sklearn.svm import SVC
import numpy as np
from tensorflow import keras
import time
from sklearn.multiclass import OneVsRestClassifier
from sklearn.svm import LinearSVC

### Dataset normalization

In [11]:
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
x_train, y_train = x_train[:30000], y_train[:30000]
x_valid, y_valid = x_test[:3000], y_test[:3000]
x_test, y_test = x_test[3000:], y_test[3000:]

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_valid = x_valid.astype('float32')

x_train = x_train / 255.0
x_test = x_test / 255.0
x_valid = x_valid / 255.0


In [12]:
x_train_flat = np.array([x.ravel() for x in x_train])
x_test_flat = np.array([x.ravel() for x in x_test])

In [15]:
start_t = time.time()
clf = SVC(gamma='auto')
clf.fit(x_train_flat, y_train)
print('Training time: ', time.time()-start_t, 'seconds')

Training time:  98.5876259803772 seconds


In [4]:
start_t = time.time()
clf = LinearSVC()
clf.fit(x_train_flat, y_train)
print('Training time: ', time.time()-start_t, 'seconds')

Training time:  0.8498804569244385 seconds


/home/fefy/anaconda3/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


In [13]:
start_t = time.time()
clf = OneVsRestClassifier(LinearSVC(), n_jobs=-1)
clf.fit(x_train_flat, y_train)
print('Training time: ', time.time()-start_t, 'seconds')

Training time:  12.749311447143555 seconds


In [17]:
start_t = time.time()
clf = OneVsRestClassifier(SVC(gamma='auto'), n_jobs=-1)
clf.fit(x_train_flat, y_train)
print('Training time: ', time.time()-start_t, 'seconds')

Training time:  207.38757920265198 seconds


In [18]:
def compute_accuracy(y_true, y_pred):
    '''Compute classification accuracy with a fixed threshold on distances.
    '''
    return np.mean(y_pred == y_true)

results_predict = clf.predict(x_test_flat)
acc = compute_accuracy(y_test, results_predict)
acc

0.9304285714285714

#### SVC - RBF - ACC: 0.9475714285714286
#### SVC - Linear - ACC: 
#### LinearSVC - ACC: 0.8784285714285714
#### OneVsRestClassifier - SVC linear - ACC: 0.8804285714285714

#### 30.000 samples
#### OneVsRestClassifier - LinearSVC - ACC: 0.9215714285714286 Training time:  12.749311447143555 seconds
#### SVC - RBF - ACC: 0.9475714285714286 Training time:  98.5876259803772 seconds
#### OneVsRestClassifier - SVC RBF - ACC: 0.9304285714285714 Training time:  207.38757920265198 seconds
